sw/host/opentitanlib/src/util/unknown.rs

```diff
@@ -0,0 +1,246 @@
+///         False = 0x14d,
+///     }
+/// }
+/// ```
+///
+/// This "enum" can be used later in match statements:
+/// ```
+/// match foo {
+///     HardenedBool::True => do_the_thing(),
+///     HardenedBool::False => do_the_opposite_thing(),
+///     HardenedBool(x) => panic!("Oh noes! {} is neither True nor False!", x),
+/// }
+/// ```
+///
+/// Behind the scenes, `with_unknown!` implements a newtype struct and
+/// creates associated constants for each of the enumerated values.
+/// The struct also implements `Copy`, `Clone`, `PartialEq`, `Eq`,
+/// `PartialOrd`, `Ord`, `Hash`, `Debug` and `Display` (including the hex,
+/// octal and binary versions).
+///
+/// In addition, `serde::Serialize` and `serde::Deserialize` are
+/// implemented.  The serialized form is a string for known values and an
+/// integer for all unknown values.
+#[macro_export]
+macro_rules! with_unknown {
+    () => {};
+
+    (
+        $(#[$outer:meta])*
+        $vis:vis enum $Enum:ident: $type:ty {
+            $(
+                $(#[$inner:meta])*
+                $enumerator:ident = $value:expr,
+            )*
+        }
+        $($more:tt)*
+    ) => {
+        $(#[$outer])*
+        #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+        #[repr(C)]
+        $vis struct $Enum(pub $type);
+
+        #[allow(non_upper_case_globals)]
+        impl $Enum {
+            $(
+                $(#[$inner])*
+                $vis const $enumerator: $Enum = $Enum($value);
+            )*
+        }
+
+        // Implement the various display traits.
+        $crate::__impl_fmt_unknown!(Display, "{}", "{}", $Enum { $($enumerator),* });
+        $crate::__impl_fmt_unknown!(LowerHex, "{:x}", "{:#x}", $Enum { $($enumerator),* });
+        $crate::__impl_fmt_unknown!(UpperHex, "{:X}", "{:#X}", $Enum { $($enumerator),* });
+        $crate::__impl_fmt_unknown!(Octal, "{:o}", "{:#o}", $Enum { $($enumerator),* });
+        $crate::__impl_fmt_unknown!(Binary, "{:b}", "{:#b}", $Enum { $($enumerator),* });
+
+        // Manually implement Serialize and Deserialize to have tight control over how
+        // the struct is serialized.
+        const _: () = {
+            use serde::ser::{Serialize, Serializer};
+            use serde::de::{Deserialize, Deserializer, Error, Visitor};
+            impl Serialize for $Enum {
+                /// Serializes the enumerated values.  All named discriminants are
+                /// serialized to strings.  All unknown values are serialized as
+                /// integers.
+                fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+                where
+                    S: Serializer,
+                {
+                    match *self {
+                        $(
+                            $Enum::$enumerator => serializer.serialize_str(stringify!($enumerator)),
+                        )*
+                        $Enum(value) => serializer.serialize_u64(value as u64),
```

Comment:
How about `serializer.serialize_u64(value.into())`? That way, it will fail to compile on types that cannot be losslessly converted to a `u64`.

Reply:
That won't work as signed types don't have an `into` for unsigned types.  I've worked up a solution with try_into.

Reply:
I see you're using `try_from` for deserialization, but I don't see your solution for serialization.

What do you want to happen if the user specifies a signed type for their enum? As-is, this will always serialized as an unsigned value.

Reply:
Even better: `serialize` is implemented on all primitive types.